core/src/main/java/org/apache/accumulo/core/security/crypto/impl/AESCryptoService.java

```diff
@@ -0,0 +1,518 @@
+    Key fek = KeyManager.unwrapKey(parsed.getEncFek(), kek);
+    switch (parsed.getCryptoServiceVersion()) {
+      case AESCBCCryptoModule.VERSION:
+        cm = new AESCBCCryptoModule(this.encryptingKek, this.encryptingKekId,
+            this.encryptingKeyManager);
+        return (cm.getDecrypter(fek));
+      case AESGCMCryptoModule.VERSION:
+        cm = new AESGCMCryptoModule(this.encryptingKek, this.encryptingKekId,
+            this.encryptingKeyManager);
+        return (cm.getDecrypter(fek));
+
+      case NO_CRYPTO_VERSION:
+        return new NoFileDecrypter();
+      default:
+        throw new CryptoException(
+            "Unknown crypto module version: " + parsed.getCryptoServiceVersion());
+    }
+  }
+
+  static class ParsedCryptoParameters {
+    String cryptoServiceName;
+    String cryptoServiceVersion;
+    String keyManagerVersion;
+    String kekId;
+    byte[] encFek;
+
+    public String getCryptoServiceName() {
+      return cryptoServiceName;
+    }
+
+    public void setCryptoServiceName(String cryptoServiceName) {
+      this.cryptoServiceName = cryptoServiceName;
+    }
+
+    public String getCryptoServiceVersion() {
+      return cryptoServiceVersion;
+    }
+
+    public void setCryptoServiceVersion(String cryptoServiceVersion) {
+      this.cryptoServiceVersion = cryptoServiceVersion;
+    }
+
+    public String getKeyManagerVersion() {
+      return keyManagerVersion;
+    }
+
+    public void setKeyManagerVersion(String keyManagerVersion) {
+      this.keyManagerVersion = keyManagerVersion;
+    }
+
+    public String getKekId() {
+      return kekId;
+    }
+
+    public void setKekId(String kekId) {
+      this.kekId = kekId;
+    }
+
+    public byte[] getEncFek() {
+      return encFek;
+    }
+
+    public void setEncFek(byte[] encFek) {
+      this.encFek = encFek;
+    }
+
+  }
+
+  private static byte[] createCryptoParameters(String version, Key encryptingKek,
+      String encryptingKekId, String encryptingKeyManager, Key fek) {
+
+    byte[] bytes;
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream params = new DataOutputStream(baos)) {
+      params.writeUTF(AESCryptoService.class.getName());
+      params.writeUTF(version);
+      params.writeUTF(encryptingKeyManager);
+      params.writeUTF(encryptingKekId);
+      byte[] wrappedFek = KeyManager.wrapKey(fek, encryptingKek);
+      params.writeInt(wrappedFek.length);
+      params.write(wrappedFek);
+
+      bytes = baos.toByteArray();
+    } catch (IOException e) {
+      throw new CryptoException("Error creating crypto params", e);
+    }
+    return bytes;
+  }
+
+  private static ParsedCryptoParameters parseCryptoParameters(byte[] parameters) {
+    ParsedCryptoParameters parsed = new ParsedCryptoParameters();
+    try (ByteArrayInputStream bais = new ByteArrayInputStream(parameters);
+        DataInputStream params = new DataInputStream(bais)) {
+      parsed.setCryptoServiceName(params.readUTF());
```

Comment:
It seems like this method is expected to handle a file created by the NoCryptoService, however looking at the code I think it would fail when given that input. Seems like the multiple calls to readUTF will fail.  Is there a test that ensures this code can read an unencrypted file?  Would be nice to have a test that compacts unencrypted files to encrypted files to run through the use case or turning on encryption on an existing system.

Reply:
I think you are right.  I will add a Test for this, good idea.

Reply:
Good eye!  I created a test to read an unencrypted RFile with an AESCryptoService and got an error on that line.  I will fix this case and push the Test with the fix.

Reply:
Do you think we also need to test with old versions of RFile like RFileTest.testOldVersions() ?